File: objectivec/day8_part2_2020.m
#import <Foundation/Foundation.h>

typedef NS_ENUM(NSInteger, OperationType) {
    acc,
    jmp,
    nop
};

@interface Instruction : NSObject
@property (nonatomic) OperationType operation;
@property (nonatomic) NSInteger argument;
@end

@implementation Instruction
- (instancetype)initWithOperation:(OperationType)operation argument:(NSInteger)argument {
    self = [super init];
    if (self) {
        _operation = operation;
        _argument = argument;
    }
    return self;
}
@end

NSInteger executeProgram(NSArray<Instruction *> *instructions, BOOL *terminated) {
    NSMutableSet<NSNumber *> *visitedIndices = [NSMutableSet set];
    NSInteger accumulator = 0;
    NSInteger currentIndex = 0;

    while (currentIndex < instructions.count) {
        if ([visitedIndices containsObject:@(currentIndex)]) {
            *terminated = NO;
            return accumulator;
        }
        [visitedIndices addObject:@(currentIndex)];

        Instruction *instruction = instructions[currentIndex];
        switch (instruction.operation) {
            case acc:
                accumulator += instruction.argument;
                currentIndex++;
                break;
            case jmp:
                currentIndex += instruction.argument;
                break;
            case nop:
                currentIndex++;
                break;
        }
    }

    *terminated = YES;
    return accumulator;
}

NSInteger fixAndExecuteProgram(NSArray<Instruction *> *instructions) {
    for (NSInteger i = 0; i < instructions.count; i++) {
        Instruction *originalInstruction = instructions[i];
        Instruction *modifiedInstruction;

        if (originalInstruction.operation == jmp) {
            modifiedInstruction = [[Instruction alloc] initWithOperation:nop argument:originalInstruction.argument];
        } else if (originalInstruction.operation == nop) {
            modifiedInstruction = [[Instruction alloc] initWithOperation:jmp argument:originalInstruction.argument];
        } else {
            continue;
        }

        NSMutableArray<Instruction *> *modifiedInstructions = [instructions mutableCopy];
        modifiedInstructions[i] = modifiedInstruction;

        BOOL terminated;
        NSInteger accumulator = executeProgram(modifiedInstructions, &terminated);
        if (terminated) {
            return accumulator;
        }
    }

    return -1; // Should never reach here
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *filePath = @"input.txt";
        NSString *fileContents = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
        NSArray<NSString *> *lines = [fileContents componentsSeparatedByString:@"\n"];

        NSMutableArray<Instruction *> *instructions = [NSMutableArray array];
        for (NSString *line in lines) {
            if (line.length == 0) continue;
            NSArray<NSString *> *components = [line componentsSeparatedByString:@" "];
            OperationType operation;
            if ([components[0] isEqualToString:@"acc"]) {
                operation = acc;
            } else if ([components[0] isEqualToString:@"jmp"]) {
                operation = jmp;
            } else if ([components[0] isEqualToString:@"nop"]) {
                operation = nop;
            } else {
                continue;
            }
            NSInteger argument = [components[1] integerValue];
            Instruction *instruction = [[Instruction alloc] initWithOperation:operation argument:argument];
            [instructions addObject:instruction];
        }

        BOOL terminated;
        NSInteger accumulatorBeforeLoop = executeProgram(instructions, &terminated);
        NSLog(@"Accumulator value before second execution: %ld", (long)accumulatorBeforeLoop);

        NSInteger accumulatorAfterFix = fixAndExecuteProgram(instructions);
        NSLog(@"Accumulator value after fixing the program: %ld", (long)accumulatorAfterFix);
    }
    return 0;
}

File: swift/day8_part2_2020.swift

import Foundation

let fileURL = URL(fileURLWithPath: "input.txt")
let input = try String(contentsOf: fileURL, encoding: .utf8)
var instructions = input.components(separatedBy: .newlines)

func runProgram(instructions: [String]) -> (Int, Bool) {
    var accumulator = 0
    var index = 0
    var visited = Set<Int>()

    while index < instructions.count {
        if visited.contains(index) {
            return (accumulator, false)
        }
        visited.insert(index)

        let parts = instructions[index].components(separatedBy: " ")
        let operation = parts[0]
        let argument = Int(parts[1])!

        switch operation {
        case "acc":
            accumulator += argument
            index += 1
        case "jmp":
            index += argument
        case "nop":
            index += 1
        default:
            break
        }
    }

    return (accumulator, true)
}

func fixProgram(instructions: [String]) -> Int {
    for i in 0..<instructions.count {
        var modifiedInstructions = instructions
        let parts = instructions[i].components(separatedBy: " ")

        if parts[0] == "jmp" {
            modifiedInstructions[i] = "nop \(parts[1])"
        } else if parts[0] == "nop" {
            modifiedInstructions[i] = "jmp \(parts[1])"
        }

        let result = runProgram(instructions: modifiedInstructions)
        if result.1 {
            return result.0
        }
    }

    return -1
}

let part1 = runProgram(instructions: instructions)
let part2 = fixProgram(instructions: instructions)

print(part1.0)
print(part2)
